### PR TITLE
Basic functionality for pcp_exporter

### PR DESCRIPTION
**Pull request by kmada97**
Add all required files to build pcp_exporter, and convert pmwebapi metrics into Prometheus metrics.

Signed-Off-By: Kalyan Mada <kalyan.red.mada@hpe.com>

### CODE REVIEW

File: .travis.yml
Comment (joehandzik) [inline .travis.yml:8]: You'll need to add a 'go get -t -v ./...' to the before_install portion to fix your travis failure

File: pcp_exporter.go
Comment (joehandzik) [inline pcp_exporter.go]: Minor nit, but I'd proably capitalize to PCP here.

File: pcp_exporter.go
Comment (joehandzik) [inline pcp_exporter.go]: Capitalize to PCP here too.

File: sources/pmwebapi.go
Comment (joehandzik) [inline sources/pmwebapi.go]: Either capitalize the P (newPcpSource) or capitalize all of PCP (newPCPSource).

File: sources/pmwebapi.go
Comment (joehandzik) [inline sources/pmwebapi.go]: Should be pcpPmwebapiSource

File: sources/pmwebapi.go
Comment (joehandzik) [inline sources/pmwebapi.go]: What is this even supposed to be doing? It's empty...

File: sources/pmwebapi.go
Comment (joehandzik) [inline sources/pmwebapi.go]: Add a space after this brace.

File: sources/pmwebapi.go
Comment (joehandzik) [inline sources/pmwebapi.go]: Rename this variable to something more descriptive than "test".

File: sources/pmwebapi.go
Comment (joehandzik) [inline sources/pmwebapi.go]: metricHeaderVar

File: sources/pmwebapi.go
Comment (joehandzik) [inline sources/pmwebapi.go]: timestampHeaderVar

File: sources/sourcy.go
Comment (joehandzik) [inline sources/sourcy.go]: Change the filename from sourcy.go to something less weird.

**Review by joehandzik — CHANGES_REQUESTED** [review]

**Comment by kmada97** [comment]
@joehandzik took care of the modifications, let me know if there is anything else!

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go:95]: Though I don't think it has to be in this PR, these `http.Get` calls are prime candidates for a new function. By my count, you can create a single function and replace three nearly identical calls in your `newPcpSource` function. You can use the following pseudo-code (haven't tested, so you might need to use your imagination):

```go
func httpRequest(url string) (string, error) {
    response, err := http.Get(url)
    if err != nil {
        return nil, err
    }
    defer response.Body.Close()
    return ioutil.ReadAll(response.Body), nil
}
```
Then you can easily just call it with 
```go
response, err := getRequest(url)
```
in `newPcpSource`. I will leave the `url` part (what it needs to be) as a future exercise for you.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go:120]: I would consider creating a single function for all of these JSON unmarshalling blocks, similar to my previous comment, but I'm not as passionate about this one since you won't be minimizing too much.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go]: Looking at this again, [log.Fatal](https://golang.org/pkg/log/#Fatal) is essentially a print followed by a non-zero system exit, so this line will never be called. I would remove it and any other return statements that are the immediate logical followup to a `log.Fatal` call.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go]: See my comment on line 116.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go]: It would be wise to do something other than just `log.Println(err)` here. The way this is currently written, if the code fails to unmarshal your `tokenBody` into `tokenMap`, it will still try and continue with the remaining code, though it will immediately fail on line 109 when it tries to access the `context` in `tokenMap`.

You can either change this to `return nil, err` or `log.Fatal(err)`. I don't have a preference, but both of these will cause your code to fail gracefully instead of something scary later on.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go]: See my comment on line 107.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go]: See my comment on line 107, though in this case, maybe it would be better to just `continue` instead of failing. That way, you can still try and collect _some_ metrics instead of failing out. Your call on whether you want to keep the error printing though, just in case you want to see that a specific metric is causing havoc.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go:152]: Something for future reference, though I think it's OK for now - maybe check that you can parse an `int` from the string _before_ using it in the struct? I know it _should_ be an `int`, but something could always go wrong.

But again, I don't see this as a big concern at the moment.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go:85]: I don't feel too strongly about this (though it would be a very easy fix), but you probably don't need to return an error. The way the code is written, it will always return `nil`, so you might as well remove the return. Note that you might have to change your call to `Update` on line 12 in `sources/source.go`.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go:39]: Not a bad idea to change these to `Seconds` and `Microseconds` (while adding
```go
`json:"s"`
```
or
```go
`json:"us"`
```
at the end). Just to make it clear what these are.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go:46]: See my comment about these names on line 39.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go:14]: You will probably want to change this to `getToken` since it is only used in this file. Otherwise, it will be exported unnecessarily and might cause issues in the future. Note that you will have to change these in the functions below where you use them. Also, it is entirely possible that a few of these need to be kept uppercase since they are embedded structs, but I am not positive on this.

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go]: Same as above: `metricHeader`

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go]: Same as above: `pcpMetrics`

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go]: Same as above: `timestampHeader`

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go]: Same as above: `timestamp`

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go]: Same as above: `getValues`

File: sources/pmwebapi.go
Comment (roclark) [inline sources/pmwebapi.go]: Same as above: `instanceList`

**Review by roclark — CHANGES_REQUESTED** [review]
I put a lot of items here, some bigger than others. Note that not all of them are critical, so you don't necessarily have to do everything I listed in this PR, but I would like to see them taken care of at some point.

The specific ones I want to see fixed before I approve are the following:
* printing/logging errors (see line 107 and linked comments)
* removing returns after `log.Fatal` calls (see line 116 and related)

Again, it would be great if you could take care of some of the others now, but I don't consider them critical.

Otherwise, great work! You are officially collecting metrics from PCP and are almost ready for your first professional merge!

File: .travis.yml
Comment (joehandzik) [reply .travis.yml:8]: This is resolved, thanks!

File: sources/pmwebapi.go
Comment (joehandzik) [reply sources/pmwebapi.go:14]: Agreed in general, though doesn't the linter/checker catch this sort of issue?

File: sources/pmwebapi.go
Comment (joehandzik) [reply sources/pmwebapi.go:39]: I like this suggestion, though would be ok with it being a separate follow-on PR.

File: sources/pmwebapi.go
Comment (joehandzik) [reply sources/pmwebapi.go:85]: Yeah, we did the same in lustre_exporter a few days ago. Another thing that can be done as a follow-on if desired, though I figure this one is even easier to do here.

File: sources/pmwebapi.go
Comment (joehandzik) [reply sources/pmwebapi.go:95]: Yup, agreed on the change, but would lean towards making it a follow-on PR.

File: sources/pmwebapi.go
Comment (joehandzik) [reply sources/pmwebapi.go]: Agreed, and have no real preference.

File: sources/pmwebapi.go
Comment (joehandzik) [reply sources/pmwebapi.go]: Agreed.

File: sources/pmwebapi.go
Comment (joehandzik) [reply sources/pmwebapi.go:120]: Definitely would make this a follow-on PR.

File: sources/pmwebapi.go
Comment (joehandzik) [reply sources/pmwebapi.go]: I would print the error but continue, just as you suggested.

File: sources/pmwebapi.go
Comment (joehandzik) [reply sources/pmwebapi.go:152]: One to worry about later for sure.

**Review by joehandzik — APPROVED** [review]

**Comment by joehandzik** [comment]
I approve, pending whatever fixes you can get done in the next day or so based on @roclark's review.

**Comment by roclark** [comment]
It looks like @joehandzik and I are pretty much on the same page - the important ones are the `log.Fatal` related lines, otherwise everything else should probably wait for future PR's.

File: sources/pmwebapi.go
Comment (roclark) [reply sources/pmwebapi.go:14]: I thought it did too... so it's odd that it didn't catch these.

Comment (joehandzik) [reply sources/pmwebapi.go:14]: I know for sure I've seen it flag me for having global structs without a header comment, so...not sure what the deal is.

Comment (joehandzik) [reply sources/pmwebapi.go:14]: Oh, hold the phone. These are used outside of this file. They're used in the unmarshalling code, that's why they need to be capitalized.

Comment (roclark) [reply sources/pmwebapi.go:14]: Great catch! That would do it! I will remove these suggestions.

File: sources/pmwebapi.go
Comment (roclark) [reply sources/pmwebapi.go:85]: Looking into this further, we will have to make several changes up the chain, including getting rid of the error checking on line 59 of `pcp_exporter.go`. I think it would be wise to keep this for now assuming we add error checking to this function when we add the more advanced label/type filtering.

**Review by roclark — APPROVED** [review]

**Comment by roclark** [comment]
The main issues I wanted to see fixed are resolved now, so once Travis passes, I am good to merge under the assumption that the other points @joehandzik and I put here will be fixed in future PRs.

File: sources/pmwebapi.go
Comment (roclark) [reply sources/pmwebapi.go:120]: Based on offline discussions, you were close in getting this to work @kmada97. I quickly hacked together an ugly example here: https://play.golang.org/p/ELpRW2cTQ9

I will leave you to figure out which parts apply to your code and what you need to change.

**Pull request merged**
